pgsm_errors.c

```diff
@@ -0,0 +1,197 @@
+#include <stdarg.h>
+#include <stdbool.h>
+#include <string.h>
+#include <time.h>
+#include <sys/time.h>
+
+#include <postgres.h>
+#include <access/hash.h>
+#include <storage/shmem.h>
+#include <utils/hsearch.h>
+
+#include "pg_stat_monitor.h"
+#include "pgsm_errors.h"
+
+
+PG_FUNCTION_INFO_V1(pg_stat_monitor_errors);
+PG_FUNCTION_INFO_V1(pg_stat_monitor_reset_errors);
+
+
+/*
+ * Maximum number of error messages tracked.
+ * This should be set to a sensible value in order to track
+ * the different type of errors that pg_stat_monitor may
+ * report, e.g. out of memory.
+ */
+#define PSGM_ERRORS_MAX 128
+
+static HTAB *pgsm_errors_ht = NULL;
+
+void psgm_errors_init(void)
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = ERROR_MSG_MAX_LEN;
+	info.entrysize = sizeof(ErrorEntry);
+	pgsm_errors_ht = ShmemInitHash("pg_stat_monitor: errors hashtable",
+									PSGM_ERRORS_MAX,  /* initial size */
+									PSGM_ERRORS_MAX,  /* maximum size */
+									&info,
+									HASH_ELEM | HASH_STRINGS);
+}
+
+size_t pgsm_errors_size(void)
+{
+    return hash_estimate_size(PSGM_ERRORS_MAX, sizeof(ErrorEntry));
+}
+
+void pgsm_log_error(const char *format, ...)
+{
+	char key[PSGM_ERRORS_MAX];
+	ErrorEntry *entry;
+	bool found = false;
+	va_list ap;
+	int n;
+	struct timeval tv;
+	struct tm *lt;
+	pgssSharedState *pgss;
+
+	va_start(ap, format);
+	n = vsnprintf(key, PSGM_ERRORS_MAX, format, ap);
+	va_end(ap);
+
+	if (n < 0)
+		return;
+
+	pgss = pgsm_get_ss();
+	LWLockAcquire(pgss->errors_lock, LW_EXCLUSIVE);
+
+	entry = (ErrorEntry *) hash_search(pgsm_errors_ht, key, HASH_ENTER_NULL, &found);
+	if (!entry)
+	{
+		/* 
+		* We're out of memory, can't track this error message.
+		* In this case we must fallback to PostgreSQL log facility.
+		*/
+		elog(WARNING, "pgsm_log_error: <OUT OF MEMORY>");
+		return;
```

Comment:
Shouldn't we `LWLockRelease` before return?
```suggestion
		LWLockRelease(pgss->errors_lock);
		return;
```

Reply:
Indeed, nice catch!